### DISCUSSION TITLE
Association table isn't auto-created in my Flask project when ORM is used. Any way I could create it with an extra API call at least?

### DISCUSSION THREAD

**Discussion by nekonyanekonya** [sqlalchemy/sqlalchemy#11673] [Usage Questions] [answered]
Hi. My base ORM works just fine, including the association object and two association proxies related to it (so, rather elaborated one). But for some reason, using the default example of building many-to-many relationship that uses basic association table has been giving me headaches, and never worked properly.

With try and error, I found declaration that at least allowed me to run the project without errors at db initialization step. But association table `assoc_tbl_1_ride_X_challenge` never was auto-created for me on the usual `db.create_all()` call, neither it was auto-created when I tried to `append()` first object that supposed to use the relationship:
```
ride.challenges.append(chal)
```
Appending like that was throwing the "table assoc_tbl_1_ride_X_challenge not exist"-kind of error all the time for that relationship, while the very same style of appending worked for my other many-to-many based on association object (table for which is created automatically).

If I create that  `assoc_tbl_1_ride_X_challenge` table manually (from MySQL's cli), then everything starts to work. So I hoped somebody at least would suggest the most ORM-ish way of creating this extra table after `db.create_all()` step. I just need it to work, and switch to developing actual features.

My declarations look about this (a lot of unrelated columns removed for clarity):

```
class Ride(db.Model):
    __tablename__ = "rideRecordingsTbl"
    id = db.Column(db.Integer, primary_key=True, autoincrement=True)
    rideName = db.Column(db.String(150))

class Challenge(db.Model):
    __tablename__ = "challengesTbl"
    id = db.Column(db.Integer, primary_key=True, autoincrement=True)
    challengeName = db.Column(db.String(150))

class Base(db.DeclarativeBase):
    pass

association_table1 = db.Table(
    "assoc_tbl_1_ride_X_challenge",
    Base.metadata,
    db.Column("rideId", db.Integer, db.ForeignKey(Ride.id), primary_key=True),
    db.Column("challengeId", db.Integer, db.ForeignKey(Challenge.id), primary_key=True)
)

Challenge.rides = db.relationship('Ride', secondary=association_table1, back_populates="challenges")
Ride.challenges = db.relationship('Challenge', secondary=association_table1, back_populates="rides")
```

**Comment by zzzeek** [comment]
make sure the module in which `association_table1` is declared is **imported** before you call `create_all()`.  that would be the main reason you see it not in the create

**Reply by nekonyanekonya** [reply]
It is. It appears in `init_db.py` script which is loosely structured like this:

```
from project import db, create_app, models
# "models" is where the ORM declarations dwell
...
# some other imports
...

def main():
# some other code
...

    app = create_app()
    if app == None:
        raise Exception("Init error, exiting..")
    with app.app_context():                                                                                                                                                                                  
        db.create_all()

# some other code
...

if __name__ == '__main__':
    main()
```

I run this file to create the db and fill it with test data, then I start the main Flask app.

**Reply by zzzeek** [reply]
do a pdb or print statement and look at metadata.tables collection, see that it's there.  look at echoed SQL output, make sure you see if checking that the table exists

**Reply by nekonyanekonya** [reply]
Okay, I'm lost..

Here is what I've tried to do:
```
    tables = []
    with app.app_context():
        print("Inside the app context...")
        db.create_all()
        metadata = MetaData()
        for table in metadata.tables:
            tables.append(table)

    for table in tables:
        print("table has type %s and contains %s"%(type(table),table))
```

And lines like "table has type ..." are not appearing in the stdout at all.. I see print's output before and after that section (the script definitely is not interrupted at that line) - but not this command's. Not sure how it's possible, haven't seen this happening in Python.

**Reply by zzzeek** [reply]
i dont see the purpose of that new metadata, if you have `db`, that's your metadata, look in there

**Reply by nekonyanekonya** [reply]
I really appreciate your help, but the problem is that I'm fairly new to all this stuff, and it's hard for me to understand how I need to check and what. It's getting fairly technical and low-level, for no good reason. It's been third day of battling it (on and off), and at that point I would rather stick in some hack that creates that one table and move on.. It's just a small community project, and my friend is waiting on me to do my part, so he could jump in too.

Any way I could just add an extra line that creates just that table to the code? I mean, I could just code in some SQL request that does it, but thought may be using some API function will be a bit better and in line with the rest of the code.

**Reply by nekonyanekonya** [reply]
That's what I got with this code:
```
       metadata = db.metadata
       print("metadata is of type %s and contains %s"%(type(metadata),metadata))
       for table in tables:
            print("table has type %s and contains %s"%(type(table),table))
```

Result:
```
metadata is of type <class 'sqlalchemy.sql.schema.MetaData'> and contains MetaData()
table has type <class 'str'> and contains usersTbl
table has type <class 'str'> and contains rideRecordingsTbl
table has type <class 'str'> and contains challengesTbl
table has type <class 'str'> and contains mileagePerchallengesTbl
```
The association table isn't in the list...

**Reply by zzzeek** [reply]
this is pretty much a flask issue that they might have better ways to help you with however take a look that "db" is the same object all the way through and there are not more than one "db" object

**Comment by CaselIT** [comment, marked as answer]
Hi, don't do this
```
class Base(db.DeclarativeBase):
    pass

association_table1 = db.Table(
    "assoc_tbl_1_ride_X_challenge",
    Base.metadata,
    ...
```

Use instead the same metadata as db
```
association_table1 = db.Table(
    "assoc_tbl_1_ride_X_challenge",
    db.metadata,
    ...
```

Sadly flash+sqlalchemy does a lot of stuff differently for no apparent reason and causes confusion

**Reply by zzzeek** [reply]
omg those are not the same two metadatas?   how unnecessary, wow

**Reply by gordthompson** [reply]
> Sadly flash+sqlalchemy does a lot of stuff differently for no apparent reason and causes confusion

FYI, David has released [Flask-SQLAlchemy-Lite](https://pypi.org/project/Flask-SQLAlchemy-Lite/) as a potential replacement for Flask-SQLAlchemy.

> Intended to be a replacement for [Flask-SQLAlchemy](https://flask-sqlalchemy.readthedocs.io/). Unlike the prior extension, this one does not attempt to manage the model base class, tables, metadata, or multiple binds for sessions. This makes the extension much simpler, letting the developer use standard SQLAlchemy instead.

**Reply by nekonyanekonya** [reply]
@CaselIT You are the savior, it works at last!
@gordthompson I don't have mental capacity left atm to rewrite it from scratch... I'll keep the new extension in mind when I'll get back to it in the future, thanks.

Thank you for your time, guys! You're the best!

**Reply by domdfcoding** [reply]
I know it isn't sqlalchemy's fault, but could a warning about this be put in the docs next to the tutorial that suggests doing it the "wrong" way for flask-sqlalchemy? I've just been bitten myself by this, and it wasn't clear why until I found this thread.

**Reply by zzzeek** [reply]
why wouldnt this warning be in flask-sqlalchemy's documentation since it's a specific pattern for that library?

**Reply by CaselIT** [reply]
agree with mike, like I wrote in the above post, this is very much flask-sqlalchemy that has decided to do a lot of stuff differently for no apparent reason, so any note or warning should be on their docs
